File: app/src/main/java/studio/mandysa/music/ui/base/BaseActivity.java
package studio.mandysa.music.ui.base;

import android.os.Bundle;

import com.yanzhenjie.sofia.Sofia;

import mandysax.fragment.FragmentActivity;
import mandysax.lifecycle.ViewModel;
import mandysax.lifecycle.ViewModelProviders;
import studio.mandysa.music.ui.viewmodel.ShareViewModel;

public abstract class BaseActivity extends FragmentActivity {
    private ShareViewModel mViewModel;

    public final <T extends ViewModel> T getViewModel(Class<T> viewmodel) {
        return ViewModelProviders.of(this).get(viewmodel);
    }

    public final ShareViewModel getShareViewModel() {
        return mViewModel == null ?
                mViewModel = ViewModelProviders.of(this, new ViewModelProviders.Factory() {
                    @Override
                    public <T extends ViewModel> T create(Class<T> modelClass) {
                        return (T) new ShareViewModel(BaseActivity.this);
                    }
                }).get(ShareViewModel.class) : mViewModel;
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        Sofia.with(this).invasionStatusBar().invasionNavigationBar();
    }

}

File: app/src/main/java/studio/mandysa/music/ui/base/BaseActivity.kt
package studio.mandysa.music.ui.base

import android.app.Activity
import android.view.LayoutInflater
import androidx.viewbinding.ViewBinding
import mandysax.fragment.FragmentActivity
import mandysax.lifecycle.*

open class BaseActivity : FragmentActivity() {
    inline fun <reified VB : ViewBinding> Activity.inflate() = lazy {
        inflateBinding<VB>(layoutInflater)
    }

    @Suppress("UNCHECKED_CAST")
    inline fun <reified VB : ViewBinding> inflateBinding(layoutInflater: LayoutInflater) =
        VB::class.java.getMethod("inflate", LayoutInflater::class.java)
            .invoke(null, layoutInflater) as VB

    inline fun <reified VB : ViewModel> ViewModelStoreOwner.viewModels() = lazy {
        return@lazy ViewModelProviders.of(
            this
        )[VB::class.java]
    }

}